File: 494. Target Sum-3.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        # p
        # s - p
        # p - (s - p) = t
        # 2p = s + t
        # p = (s + t) / 2
        target += sum(nums)
        if target < 0 or target % 2:
            return 0
        target //= 2

        n = len(nums)

        f = [[0] * (target + 1) for _ in range(2)]
        f[0][0] = 1

        for i, x in enumerate(nums):
            for c in range(target + 1):
                if c < x:
                    f[(i + 1) % 2][c] = f[i % 2][c]
                else:
                    f[(i + 1) % 2][c] = f[i % 2][c] + f[i % 2][c - x]

        return f[n % 2][target]

File: 494. Target Sum-3.cpp
class Solution {
public:
  int findTargetSumWays(vector<int> &nums, int target) {
    int sum = accumulate(nums.begin(), nums.end(), 0);
    if (sum < abs(target) || (sum + target) % 2 != 0)
      return 0;
    int P = (sum + target) / 2;
    vector<int> f(P + 1, 0);
    f[0] = 1;
    for (auto x : nums)
      for (int i = P; i >= x; --i)
        f[i] += f[i - x];
    return f[P];
  }
};